### PR TITLE
Allow filtering Python packages hosted on Pulp

### PR DESCRIPTION
**Pull request by fridex**
## This introduces a breaking change

- [x] No

### CODE REVIEW

File: thoth/adviser/sieves/_pulp/no_index_label.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class NoPulpIndexLabelSieve(PulpIndexLabelSieve):
+    """A sieve that filters out packages that come from Operate First Pulp instance."""
+
+    @classmethod
+    def should_include(cls, builder_context: "PipelineBuilderContext") -> Generator[Dict[Any, Any], None, None]:
+        """Include this wrap in adviser, if label pulp-index=off is used."""
```
Comment (mayaCostantini) [inline thoth/adviser/sieves/_pulp/no_index_label.py:40]:
```suggestion
        """Include this wrap in adviser, if label pulp-index=disabled is used."""
```

File: tests/sieves/_pulp/test_no_index_label.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Test a sieve that provides only packages available on Operate First Pulp instance."""
```
Comment (mayaCostantini) [inline tests/sieves/_pulp/test_no_index_label.py:18]:
```suggestion
"""Test a sieve that provides only packages available on PyPI instance."""
```

File: tests/sieves/_pulp/test_no_index_label.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+from thoth.adviser.sieves import NoPulpIndexLabelSieve
+from thoth.adviser.sieves._pulp import PULP_URL
+from thoth.python import PackageVersion
+from thoth.python import Source
+
+from ...base import AdviserUnitTestCase
+
+
+class TestNoPulpIndexLabelSieve(AdviserUnitTestCase):
+    """Test a sieve that provides only packages available on Operate First Pulp instance."""
```
Comment (mayaCostantini) [inline tests/sieves/_pulp/test_no_index_label.py:32]:
```suggestion
    """Test a sieve that provides only packages available on PyPI instance."""
```

File: tests/sieves/_pulp/test_no_index_label.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        )
+        self.verify_multiple_should_include(builder_context)
+
+    def test_no_include(self, builder_context: PipelineBuilderContext) -> None:
+        """Test not including this pipeline unit if the index=no-pulp label is not provided."""
+        builder_context.labels.clear()
+        self.UNIT_TESTED.should_include(builder_context)
+
+    def test_run(self, context: Context) -> None:
+        """Test filtering packages that are not hosted on Operate First Pulp."""
```
Comment (mayaCostantini) [inline tests/sieves/_pulp/test_no_index_label.py:49]:
```suggestion
        """Test filtering packages that are not hosted on PyPI."""
```

**Review by mayaCostantini — COMMENTED** [review]
Just a few minor things, otherwise lgtm :100:

File: tests/sieves/_pulp/test_no_index_label.py (hunk 2 above)
Comment (fridex) [reply tests/sieves/_pulp/test_no_index_label.py:18]: This is specific to Pulp instance (an alternative to PyPI that is being provisioned). See https://www.operate-first.cloud/community-handbook/pulp/usage.md

File: tests/sieves/_pulp/test_no_index_label.py (hunk 3 above)
Comment (fridex) [reply tests/sieves/_pulp/test_no_index_label.py:32]: See comment above.

File: thoth/adviser/sieves/_pulp/index_label.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+_LOGGER = logging.getLogger(__name__)
+
+
+class PulpIndexLabelSieve(Sieve):
+    """A sieve that filters out packages that do not come from Operate First Pulp instance."""
+
+    _stack_info_reported = attr.ib(type=bool, default=False, init=False)
+
+    @classmethod
+    def should_include(cls, builder_context: "PipelineBuilderContext") -> Generator[Dict[Any, Any], None, None]:
```
[the remark this answers is not in the document]
Comment (fridex) [reply thoth/adviser/sieves/_pulp/index_label.py:44]: `yield from ()` will yield an empty iterator (so [the pipeline unit is not included](https://thoth-station.ninja/docs/developers/adviser/unit.html#registering-a-pipeline-unit-to-pipeline))

File: tests/sieves/_pulp/test_no_index_label.py (hunk 2 above)
Comment (mayaCostantini) [reply tests/sieves/_pulp/test_no_index_label.py:18]: Ok I see, thanks!

**Comment by mayaCostantini** [comment]
lgtm :100: 
/approve

**Comment by fridex** [comment]
All green. Merging to continue with a new release.

**Pull request merged**